### PR TITLE
Added --tests-file and --failed-tests-file options

### PR DESCRIPTION
**Pull request by rafie**
`--tests-file` (repeatable) allows for specification of a file containing a list of test specs (`file:test`)
`--failed-tests-file` enables writing of failed tests names into a specified file

Also made `--test` repeatable.

### CODE REVIEW

File: RLTest/__main__.py [hunk 1]
```diff
@@ -171,7 +171,13 @@ def do_normal_conn(self, line):
     help='stop before each test allow gdb attachment')
 
 parser.add_argument(
-    '-t', '--test', help='Specify test to run, in the form of "file:test"')
+    '-t', '--test', action='append', help='test to run, in the form of "file:test"')
+
+parser.add_argument(
+    '-f', '--tests-file', action='append', help='file containing test to run, in the form of "file:test"')
```
Comment (chayim) [inline RLTest/__main__.py:177]: Can you see if this parser supports ```metavar=FILE```

File: RLTest/__main__.py [hunk 2]
```diff
@@ -171,7 +171,13 @@ def do_normal_conn(self, line):
...
+
+parser.add_argument(
+    '-F', '--failed-tests-file', help='destination file for failed tests')
```
Comment (chayim) [inline RLTest/__main__.py:180]: Can you see if this parser supports ```metavar=FILE```

File: RLTest/__main__.py [hunk 3]
```diff
@@ -417,7 +423,14 @@ def __init__(self):
...
             self.loader.load_spec(self.args.test)
-        else:
+        if self.args.tests_file:
+            for f in self.args.tests_file:
+                with open(f, 'r') as file:
+                    for line in file.readlines():
+                        if line.startswith('#') or line.strip() == "":
+                            continue
+                        self.loader.load_spec(line.strip())
+        if not self.args.test and not self.args.tests_file:
```
Comment (chayim) [inline RLTest/__main__.py:433]: preferred python syntax now is more explicit
```
if self.args.test is not None and self.args.tests_file is not None:
```

File: RLTest/__main__.py [hunk 4]
```diff
@@ -417,7 +423,14 @@ def __init__(self):
         self.loader = TestLoader()
         if self.args.test:
             self.loader.load_spec(self.args.test)
-        else:
+        if self.args.tests_file:
+            for f in self.args.tests_file:
+                with open(f, 'r') as file:
```
Comment (chayim) [inline RLTest/__main__.py:428]:
```
if self.args.test_file is not None and os.path.exists(self.args.test_file):
    for fname in self.args.tests_file:
        with open(fname) as fp:
            [self.loader.load_spec(line.strip()) for line in fp.readlines() if line.strip() not in ['#', '']]
```

File: RLTest/__main__.py [hunk 5]
```diff
@@ -721,6 +739,10 @@ def run_jobs(jobs, results, port):
                 for failure in failures:
                     print('\t\t' + failure)
             sys.exit(1)
+        else:
+            if self.args.failed_tests_file:
+                with open(self.args.failed_tests_file, 'w') as file:
+                    file.write('')
```
Comment (chayim) [inline RLTest/__main__.py:745]: Did you mean to write empty here? If so, either of the following are common. WDYT

```
fp = open(self.args.failed_tests_file, 'w')
fp.close()
```

or

```
with open(self.args.failed_tests_file, 'w') as file:
    pass
```

File: RLTest/loader.py [hunk 6]
```diff
@@ -3,6 +3,7 @@
 import sys
 import imp
 import inspect
+from RLTest.utils import Colors
```
Comment (chayim) [inline RLTest/loader.py:6]: IMHO We should really remove our on colors and just use the package that does this.

File: RLTest/loader.py [hunk 7]
```diff
@@ -70,12 +71,16 @@ def get_functions(self, instance):
 
 
 class TestLoader(object):
-    def __init__(self, filter=None):
+    def __init__(self):
         self.tests = []
-        self.toplevel_filter = filter
-        self.subfilter = None
 
     def load_spec(self, arg):
```
Comment (chayim) [inline RLTest/loader.py:77]: If arg can now take multiple types, we should add a type hint here (specifically).

File: RLTest/loader.py [hunk 8]
```diff
@@ -96,48 +101,53 @@ def load_spec(self, arg):
             sys.path.append(dirname)
 
         module_name, _ = os.path.splitext(os.path.basename(filename))
+        toplevel_filter = None
+        subfilter = None
         if varname:
```
Comment (chayim) [inline RLTest/loader.py:106]: Exception throwing is cheap (preferred even) in python.

```
if varname is not None:
    try:
        toplevel_filter, subfilter = varname.split('.')
    except ValueError:
        toplevel_filter=  varname
```

File: RLTest/loader.py [hunk 9]
```diff
@@ -96,48 +101,53 @@ def load_spec(self, arg):
...
-                continue
-
-            obj = getattr(module, symbol)
-            if inspect.isclass(obj):
-                methnames = [mname for mname in dir(obj)
-                             if self.filter_method(mname)]
-                self.tests.append(TestClass(filename, symbol, module_name, methnames))
-            elif inspect.isfunction(obj):
-                self.tests.append(TestFunction(filename, symbol, module_name))
+        try:
```
Comment (chayim) [inline RLTest/loader.py:116]: The try/catch isn't granular enough

File: RLTest/loader.py [hunk 10]
```diff
@@ -96,48 +101,53 @@ def load_spec(self, arg):
...
+
+                obj = getattr(module, symbol)
+                if inspect.isclass(obj):
+                    methnames = [mname for mname in dir(obj)
+                                 if self.filter_method(mname, subfilter)]
+                    self.tests.append(TestClass(filename, symbol, module_name, methnames))
+                elif inspect.isfunction(obj):
+                    self.tests.append(TestFunction(filename, symbol, module_name))
+        except:
+            print(Colors.Bred("File %s not found: skipping" % filename))
```
Comment (chayim) [inline RLTest/loader.py:132]: If this is the only error, consider doing the file check above.

**Review by chayim — CHANGES_REQUESTED** [review]
Added comments

File: RLTest/__main__.py (hunk 1 above)
Comment (rafie) [reply RLTest/__main__.py:177]: This is just for aesthetics of help message. Tested in with RedisGraph, looks ok.

File: RLTest/loader.py (hunk 7 above)
Comment (rafie) [reply RLTest/loader.py:77]: arg is never used - this is dead code.

File: RLTest/__main__.py (hunk 5 above)
Comment (rafie) [reply RLTest/__main__.py:745]: Although opened with 'w', it seems that if nothing is being written to the file, it's not being overwritten.

File: RLTest/loader.py (hunk 9 above)
Comment (rafie) [reply RLTest/loader.py:116]: This is problematic from another respect: we might create too much noise here. Maybe it's better to fail silently. Need to think.

File: RLTest/__main__.py (hunk 4 above)
Comment (rafie) [reply RLTest/__main__.py:428]: `test-file` is actually tests_files - this is a list of files.
IMO the comprehension here makes it hard to understand the code.
Added exception handling.

File: RLTest/loader.py (hunk 8 above)
Comment (rafie) [reply RLTest/loader.py:106]: except that split() doesn't throw.

Comment (chayim) [reply RLTest/loader.py:106]: Pasted the wrong bit:
```

Comment (chayim) [reply RLTest/loader.py:106]: Pasted the wrong piece - sorry:
```
try:
    toplevel_filter = varname.split('.')[0]
    subfilter = varname.split('.')[1]
except IndexError:
    toplevel_filter=  varname
except AttributeError:  # None case
    pass
```

^^ or similar.

Comment (rafie) [reply RLTest/loader.py:106]: Note that `"abc".split('.') == ['abc']` - no exception here.

But anyways, which one looks clearer?
```
    toplevel_filter, subfilter  = None, None
    if varname:
        if '.' in varname:
            toplevel_filter, subfilter = varname.split('.')
        else:
            toplevel_filter = varname
```
or
```
try:
    x = varname.split('.')
    toplevel_filter = x[0]
    subfilter = x[1]
except IndexError:
    toplevel_filter = varname
except AttributeError:  # None case
    pass
```

File: RLTest/loader.py (hunk 6 above)
Comment (rafie) [reply RLTest/loader.py:6]: It's a bit out of context for this PR, but yes, we should.

**Review by chayim — APPROVED** [review]

**Pull request merged**